File: Assignment7/statistics.ml
(* Statistics Calculator - OCaml Functional Implementation *)

(* Compute arithmetic mean using functional approach *)
let compute_average numbers =
  let total = List.fold_left (fun accumulator element -> accumulator +. (float_of_int element)) 0.0 numbers in
  let length = float_of_int (List.length numbers) in
  total /. length

(* Determine middle value in sorted list *)
let determine_middle_value numbers =
  let ordered = List.sort compare numbers in
  let size = List.length ordered in
  if size mod 2 = 0 then
    let first_middle = List.nth ordered ((size / 2) - 1) in
    let second_middle = List.nth ordered (size / 2) in
    (float_of_int (first_middle + second_middle)) /. 2.0
  else
    float_of_int (List.nth ordered (size / 2))

(* Count occurrences of each element *)
let count_occurrences numbers =
  let rec occurrence_helper result = function
    | [] -> result
    | current :: remaining ->
      let occurrence_count = List.fold_left (fun counter element -> if element = current then counter + 1 else counter) 0 numbers in
      if List.mem_assoc current result then
        occurrence_helper result remaining
      else
        occurrence_helper ((current, occurrence_count) :: result) remaining
  in
  occurrence_helper [] numbers

(* Find highest occurrence count *)
let find_highest_count occurrence_list =
  List.fold_left (fun highest (_, count) -> max highest count) 0 occurrence_list

(* Identify most frequent values *)
let identify_most_frequent numbers =
  let occurrence_list = count_occurrences numbers in
  let highest_count = find_highest_count occurrence_list in
  let frequent_values = List.filter_map (fun (value, count) -> 
    if count = highest_count then Some value else None) occurrence_list in
  List.sort compare frequent_values

(* Function to print list *)
let print_list lst =
  let rec print_helper = function
    | [] -> print_string "[]"
    | [x] -> print_string ("[" ^ string_of_int x ^ "]")
    | x :: xs -> 
      print_string ("[" ^ string_of_int x);
      List.iter (fun y -> print_string (", " ^ string_of_int y)) xs;
      print_string "]"
  in
  print_helper lst

(* Function to print float list *)
let print_float_list lst =
  let rec print_helper = function
    | [] -> print_string "[]"
    | [x] -> print_string ("[" ^ string_of_float x ^ "]")
    | x :: xs -> 
      print_string ("[" ^ string_of_float x);
      List.iter (fun y -> print_string (", " ^ string_of_float y)) xs;
      print_string "]"
  in
  print_helper lst

(* Main function *)
let main () =
  print_endline "=== Statistics Calculator (OCaml - Functional) ===";
  print_newline ();
  
  (* Test data *)
  let data = [1; 2; 3; 4; 5; 2; 3; 2; 1; 3] in
  
  print_string "Input data: ";
  print_list data;
  print_newline ();
  print_newline ();
  
  (* Calculate and display mean *)
  let mean = compute_average data in
  print_string "Mean: ";
  print_float mean;
  print_newline ();
  
  (* Calculate and display median *)
  let median = determine_middle_value data in
  print_string "Median: ";
  print_float median;
  print_newline ();
  
  (* Calculate and display mode *)
  let modes = identify_most_frequent data in
  print_string "Mode: ";
  print_list modes;
  print_newline ();
  print_newline ()

(* Run the main function *)
let () = main ()

File: Assignment7/statistics.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Compute arithmetic average
double compute_average(int numbers[], int size) {
    double total = 0.0;
    for (int i = 0; i < size; i++) {
        total += numbers[i];
    }
    return total / size;
}

// Arrange elements in ascending order
void arrange_ascending(int numbers[], int size) {
    for (int i = 0; i < size - 1; i++) {
        for (int j = 0; j < size - i - 1; j++) {
            if (numbers[j] > numbers[j + 1]) {
                int swap = numbers[j];
                numbers[j] = numbers[j + 1];
                numbers[j + 1] = swap;
            }
        }
    }
}

// Find middle value
double find_middle_value(int numbers[], int size) {
    int *temp_array = malloc(size * sizeof(int));
    memcpy(temp_array, numbers, size * sizeof(int));
    
    arrange_ascending(temp_array, size);
    
    double result;
    if (size % 2 == 0) {
        result = (temp_array[size/2 - 1] + temp_array[size/2]) / 2.0;
    } else {
        result = temp_array[size/2];
    }
    
    free(temp_array);
    return result;
}

// Determine most frequent values
void find_most_frequent(int numbers[], int size, int frequent_values[], int *count) {
    int highest_frequency = 0;
    *count = 0;
    
    for (int i = 0; i < size; i++) {
        int frequency = 0;
        for (int j = 0; j < size; j++) {
            if (numbers[j] == numbers[i]) {
                frequency++;
            }
        }
        
        if (frequency > highest_frequency) {
            highest_frequency = frequency;
            *count = 0;
            frequent_values[*count] = numbers[i];
            (*count)++;
        } else if (frequency == highest_frequency) {
            int already_exists = 0;
            for (int k = 0; k < *count; k++) {
                if (frequent_values[k] == numbers[i]) {
                    already_exists = 1;
                    break;
                }
            }
            if (!already_exists) {
                frequent_values[*count] = numbers[i];
                (*count)++;
            }
        }
    }
}

// Function to print array
void print_array(int arr[], int n) {
    printf("[");
    for (int i = 0; i < n; i++) {
        printf("%d", arr[i]);
        if (i < n - 1) printf(", ");
    }
    printf("]");
}

int main() {
    printf("=== Statistics Calculator (C - Procedural) ===\n\n");
    
    // Test data
    int data[] = {1, 2, 3, 4, 5, 2, 3, 2, 1, 3};
    int n = sizeof(data) / sizeof(data[0]);
    
    printf("Input data: ");
    print_array(data, n);
    printf("\n\n");
    
    // Calculate and display mean
    double mean = compute_average(data, n);
    printf("Mean: %.2f\n", mean);
    
    // Calculate and display median
    double median = find_middle_value(data, n);
    printf("Median: %.2f\n", median);
    
    // Calculate and display mode
    int modes[100]; // Assuming max 100 different modes
    int mode_count;
    find_most_frequent(data, n, modes, &mode_count);
    
    printf("Mode: ");
    if (mode_count == 0) {
        printf("No mode");
    } else {
        print_array(modes, mode_count);
    }
    printf("\n");
    
    return 0;
}
